---
title: "Plotting manuscript figures"
output: html_notebook
---

```{r setup}
library(tidyverse)
library(openxlsx)
library(scater)

# Path to project folder
wd <- here::here()
source(file.path(wd, "scRNAseq/analysis/scripts/colors.R"))

figures.out.dir <- file.path(wd, "scRNAseq/figures")
tables.out.dir <- file.path(wd, "scRNAseq/analysis/results")

add.flag <- function(pheatmap,
                     kept.labels,
                     repel.degree) {

  # repel.degree = number within [0, 1], which controls how much 
  #                space to allocate for repelling labels.
  ## repel.degree = 0: spread out labels over existing range of kept labels
  ## repel.degree = 1: spread out labels over the full y-axis

  heatmap <- pheatmap$gtable

  new.label <- heatmap$grobs[[which(heatmap$layout$name == "row_names")]] 

  # keep only labels in kept.labels, replace the rest with ""
  new.label$label <- ifelse(new.label$label %in% kept.labels, 
                            new.label$label, "")

  # calculate evenly spaced out y-axis positions
  repelled.y <- function(d, d.select, k = repel.degree){
    # d = vector of distances for labels
    # d.select = vector of T/F for which labels are significant

    # recursive function to get current label positions
    # (note the unit is "npc" for all components of each distance)
    strip.npc <- function(dd){
      if(!"unit.arithmetic" %in% class(dd)) {
        return(as.numeric(dd))
      }

      d1 <- strip.npc(dd$arg1)
      d2 <- strip.npc(dd$arg2)
      fn <- dd$fname
      return(lazyeval::lazy_eval(paste(d1, fn, d2)))
    }

    full.range <- sapply(seq_along(d), function(i) strip.npc(d[i]))
    selected.range <- sapply(seq_along(d[d.select]), function(i) strip.npc(d[d.select][i]))

    return(unit(seq(from = max(selected.range) + k*(max(full.range) - max(selected.range)),
                    to = min(selected.range) - k*(min(selected.range) - min(full.range)), 
                    length.out = sum(d.select)), 
                "npc"))
  }
  new.y.positions <- repelled.y(new.label$y,
                                d.select = new.label$label != "")
  new.flag <- grid::segmentsGrob(x0 = new.label$x,
                           x1 = new.label$x + unit(0.15, "npc"),
                           y0 = new.label$y[new.label$label != ""],
                           y1 = new.y.positions)

  # shift position for selected labels
  new.label$x <- new.label$x + unit(0.2, "npc")
  new.label$y[new.label$label != ""] <- new.y.positions

  # add flag to heatmap
  heatmap <- gtable::gtable_add_grob(x = heatmap,
                                   grobs = new.flag,
                                   t = 4, 
                                   l = 4
  )

  # replace label positions in heatmap
  heatmap$grobs[[which(heatmap$layout$name == "row_names")]] <- new.label

  # plot result
  grid::grid.newpage()
  grid::grid.draw(heatmap)

  # return a copy of the heatmap invisibly
  invisible(heatmap)
}
```

# From the total dataset

```{r}
iteration_input = "4_all_cells_labbeled"
sce.dir <- file.path(wd, "scRNAseq/analysis/outputs", iteration_input)
sce.total.labelled <- readRDS(file = file.path(sce.dir, "sce.rds"))
sce.total.labelled
```

## Figure 1 - Cells tSNE embedding

```{r fig.height=6, fig.width=7, message=FALSE}
Fig.1A <- plotReducedDim(
    sce.total.labelled,
    dimred = "TSNE",
    color_by = "sample_norep_plot",
    point_size = 1.5
) + 
  theme(text = element_text(size = 14),
        # plot.margin = unit(c(5.5, 200, 5.5, 5.5), "pt")
        ) +
  scale_color_manual(values = get_palettes("tableau20")) +
  labs(color = "Sample")
Fig.1A
```

```{r fig.height=5.5, fig.width=6.5, message=FALSE}
sce.total.labelled$cell_type_manual_2 <- colData(sce.total.labelled) %>%
  as.data.frame() %>%
  mutate(
    cell_type_manual = case_when(
      cell_type_manual == "T cell" &
        cell_type_fine == "Natural killer cells" ~ "NK cell",
      
      cell_type_manual == "T cell" &
        (
          str_detect(cell_type_fine, "CD4") |
            str_detect(cell_type_fine, "Th")
        )  ~ "CD4+ T cell",
      
      cell_type_manual == "T cell" &
        str_detect(cell_type_fine, "CD8")  ~ "CD8+ T cell",
      
      cell_type_manual == "B cell" &
        clusters_louvain_K30 == 11 ~ "Naive B cell",
      cell_type_manual == "MCL" ~ "MCL cell",
      TRUE ~ cell_type_manual
    ),
    cell_type_manual = factor(
      cell_type_manual,
      levels = c(
        "B cell",
        "Naive B cell",
        "T cell",
        "NK cell",
        "CD8+ T cell",
        "CD4+ T cell",
        "Monocyte",
        "MCL cell"
      )
    )
  ) %>% pull(cell_type_manual)

Fig.1B <- plotReducedDim(
  sce.total.labelled,
  dimred = "TSNE",
  color_by = "cell_type_manual_2",
  point_size = 2
) +
  scale_color_manual(
    values = c(
      "B cell" = "#006BA4",
      "Naive B cell" = "#18A188",
      "MCL cell" = "#D82526",
      "Monocyte" = "#AD8BC9",
      "T cell" = "#898989",
      "NK cell" = "#67BF5C",
      "CD4+ T cell" = "#C85200",
      "CD8+ T cell" = "#6DCCDA"
    )
  ) + labs(color = "Cell type") +
  theme(text = element_text(size = 14))
Fig.1B
```

```{r fig.height=4.75, fig.width=5, message=FALSE}
Fig.1C <- plotReducedDim(
    sce.total.labelled,
    dimred = "TSNE",
    color_by = "CCND1",
    point_size = 1.5
) + theme(text = element_text(size = 14))
Fig.1C
```

```{r, fig.height=4, fig.width=7.5}
cell.types <- c("MCL cells", "B cells", "Other normal cells")

sce.total.labelled$cell_type_manual_2 <- colData(sce.total.labelled) %>%
  as.data.frame() %>%
  mutate(
    cell_type_manual = case_when(
      cell_type_manual == "MCL" ~ "MCL cells",
      cell_type_manual == "B cell" ~ "B cells",
      TRUE ~ "Other normal cells"
    ),
    cell_type_manual = factor(
      cell_type_manual,
      levels = c(
        "MCL cells",
        "B cells",
        "Other normal cells"
      )
    )
  ) %>% pull(cell_type_manual)
cell_colors <- c(
      "B cells" = "#006BA4",
      "MCL cells" = "#D82526",
      "Other normal cells" = "#C87000"
    )

Fig.1D <- data.frame(
  Patient = sce.total.labelled$patient,
  Cell_type = sce.total.labelled$cell_type_manual_2,
  "logcounts.CCND1" = logcounts(sce.total.labelled)["CCND1", ]
) %>%
  filter(Cell_type %in% cell.types) %>%
  filter(!(Patient == "P022" & Cell_type == "B cells")) %>%
  mutate(Cell_type = factor(Cell_type, levels = cell.types)) %>%
  ggplot() +
  geom_density_ridges(
    aes(x = logcounts.CCND1, y = Cell_type, fill = Cell_type),
    panel_scaling = T,
    scale = 2,
    alpha = 0.5,
    #jittered_points = TRUE,
    #position = position_points_jitter(width = 0.05, height = 0),
    #point_shape = '|', point_size = 1, point_alpha = 0.5
    ) +
  scale_y_discrete(expand = expansion(mult = c(0.05, 1.1))) +
  theme_bw() +
  scale_fill_manual(values = cell_colors) +
  facet_wrap( ~ Patient, ncol = 5) +
  labs(y = NULL, x = "log(counts) CCND1") +
  theme(legend.position = "none", 
        text = element_text(size = 14)
        )
Fig.1D
```

```{r}
tiff(
  filename = file.path(figures.out.dir, "Fig1A.tiff"),
  units = "in",
  width = 7.5,
  height = 6,
  res = 300,
  bg = "white"
)
Fig.1A
dev.off()

tiff(
  filename = file.path(figures.out.dir, "Fig1B.tiff"),
  units = "in",
  width = 6.5,
  height = 6,
  res = 300,
  bg = "white"
)
Fig.1B
dev.off()

tiff(
  filename = file.path(figures.out.dir, "Fig1C.tiff"),
  units = "in",
  width = 5,
  height = 4.75,
  res = 300,
  bg = "white"
)
Fig.1C
dev.off()

tiff(
  filename = file.path(figures.out.dir, "Fig1D.tiff"),
  units = "in",
  width = 7,
  height = 5,
  res = 300,
  bg = "white"
)
Fig.1D
dev.off()
```

```{r fig.height=6, fig.width=10, message=FALSE}
plotReducedDim(
    sce.total.labelled,
    dimred = "TSNE",
    color_by = "cell_type_fine",
    point_size = 2
  ) +
  scale_color_manual(values = c(get_palettes("bluered12"), get_palettes("tableau10"), get_palettes("tableau20")))
```

# From the MCL dataset

List all sce directories:

```{r}
list.dirs(file.path(wd, "scRNAseq/analysis/outputs"), recursive = F, full.names = F)
```

Load:

```{r}
iteration_input = "6_MCL_CNV"
sce.dir <- file.path(wd, "scRNAseq/analysis/outputs", iteration_input)
sce.MCL <- readRDS(file = file.path(sce.dir, "sce.rds"))
sce.MCL$Sample <- sce.MCL$sample_norep_plot
sce.MCL$Sample <- factor(sce.MCL$Sample,
                         levels = levels(sce.MCL$sample_norep) %>% 
                           str_replace_all(., "_", " "))

sce.MCL
```

## Supplementary Figure 2 - MCL cells tSNE

```{r, fig.width=8.2, fig.height=4, message=FALSE, warning=FALSE}
SF3 <- plotReducedDim(sce.MCL,
               dimred = "TSNE_corrected_SamplePhase",
               color_by = "Sample",
               other_fields = "timepoint") +
  facet_wrap( ~ timepoint) +
  scale_color_manual(values = get_palettes("tableau20")) +
  labs(color = "Sample", x = "t-SNE 1", y = "t-SNE 2") +
  xlim(-35, 40) + ylim(-40, 35) +
  guides(color = guide_legend(override.aes = list(size = 4))) +
  cowplot::theme_half_open() +
  cowplot::panel_border(color = "grey30", size = 1.5) +
  theme(strip.text = element_text(face = "bold"))
SF3
```

Labelled sub-clones in the t-SNE space.

```{r}
tiff(
  filename = file.path(figures.out.dir, "SFig3.tiff"),
  units = "in",
  width = 8.2,
  height = 4,
  res = 300,
  bg = "white"
)
SF3
dev.off()
```

## Figure 3A - MCL tSNE by subclone

```{r fig.height=5, fig.width=10, message=FALSE}
fig3.A <- plotReducedDim(
  sce.MCL[, sce.MCL$patient != "P069"],
  dimred = "TSNE_corrected_SamplePhase",
  color_by = "CNV.clone",
  point_size = 1.5,
  other_fields = c("patient", "timepoint")
) +
  facet_grid(timepoint ~ patient) +
  xlim(-35, 40) + ylim(-40, 35) +
  guides(color = guide_legend(override.aes = list(size = 4))) +
  cowplot::theme_half_open(font_size = 16) + 
  cowplot::panel_border(color = "gray30", size = 1.5) +
  theme(strip.text = element_text(face = "bold")) +
  scale_color_manual(values = c("gray50", "red", "orange")) +
  labs(color = "CNV clone", x = "TSNE 1", y = "TSNE 2")
fig3.A
```

## Figure 3B and suppl. figures 3-6 

### Gene expression heatmap & GSA dotplots

```{r message=FALSE, fig.width=5, fig.height=9}
# load the Diff. Expr. tables for DGs-DGex comparison
DE.output.path <-
  file.path(wd, "scRNAseq/analysis/outputs/DE_DECENT_clone")
clone.de.list <- list()
samples <- levels(sce.MCL$patient)
for (samplename in samples) {
  filename <- paste0("DE_clone_", samplename, ".csv")
  clone.de.list[[samplename]] <-
    read_csv(file.path(DE.output.path, samplename, filename))
}

sig.pvalue = 0.01
sig.logfc = 2

clone.de.list.sig <- clone.de.list %>%
  lapply(., filter,
         abs(logfc) > sig.logfc & pvalue < sig.pvalue)

clone.de.list.sig
```

```{r}
DECENT.GSE.list <-
  readRDS(file.path(wd, "scRNAseq/analysis/results/DE_analysis_DK", "DECENT.GSE.RPA.rds"))[1:4]

lapply(DECENT.GSE.list, as.data.frame)
```

### P087 (showcase)

```{r}
current_patient = "P087"
```


```{r message=FALSE, fig.width=5, fig.height=9}
genes.to.plot <- clone.de.list.sig[[current_patient]] %>%
  arrange(desc(abs(stat))) %>%
  pull(gene) %>% 
  str_subset(., ".+-AS[0-9]", negate = T) %>% # clear novel transcripts
  str_subset(., "^LINC[0-9]+", negate = T) %>% 
  str_subset(., "^A(C|L|P)[0-9]+", negate = T)
genes.to.plot
```

```{r}
highlight.genes <- c("BCL2L10", "CCL3", "CD84", "CCL22", "CD70", "CD9", "FOS", "MAPK12", "BIK", "BCL6", "TXNIP")
```

```{r, fig.width=5, fig.height=6}
gene.exp.HT <- sce.MCL[, sce.MCL$patient == current_patient] %>%
  plotGroupedHeatmap(
    .,
    features = genes.to.plot,
    group = "CNV.clone",
    center = TRUE,
    zlim = c(-0.1, 0.1),
    cluster_rows = T,
    angle_col = c("0"),
    silent = T
  ) %>%
  add.flag(.,
           kept.labels = highlight.genes,
           repel.degree = 0.2) %>% 
  ggplotify::as.ggplot()
```

```{r}
lapply(DECENT.GSE.list, as.data.frame)[[current_patient]]
```

```{r message=FALSE}
categories.plot <- c("NIK-->noncanonical NF-kB signaling",
                     "Activation of NF-kappaB in B cells",
                     "Dectin-1 mediated noncanonical NF-kB signaling",
                     "The citric acid (TCA) cycle and respiratory electron transport",
                     "FCERI mediated NF-kB activation",
                     "Regulation of PTEN stability and activity",
                     "p53-Dependent G1/S DNA damage checkpoint",
                     "TNFR2 non-canonical NF-kB pathway",
                     "Activation of ATR in response to replication stress"
                     )

dotplot.GSE <- DECENT.GSE.list[[current_patient]] %>% 
  ReactomePA::dotplot(., showCategory = categories.plot) +
  scale_x_continuous(labels = scales::percent, limits = c(0.3, 0.9), name = "Gene ratio") +
  geom_point(aes(fill = NES)) +
  scale_fill_gradient2(name = "NES", low = "blue", high = "red")
dotplot.GSE
```

```{r, fig.width=10, fig.height=12}
fig3 <- ggpubr::ggarrange(
  fig3.A,
  ggpubr::ggarrange(gene.exp.HT,
                    dotplot.GSE, 
                    nrow = 1, labels = c("B", "C")),
  nrow = 2, heights = c(0.55, 1), labels = c("A")
)
fig3
```

```{r}
tiff(
  filename = file.path(figures.out.dir, "Fig3_DE.tiff"),
  units = "in",
  width = 10,
  height = 12,
  res = 300,
  bg = "white"
)
fig3
dev.off()
```

### Plots for all patients

```{r}
patients <- names(clone.de.list.sig)
```


```{r}
genes.list <- list()
HT.list <- list()
for (current_patient in patients) {
  genes.list[[current_patient]] <-
    clone.de.list.sig[[current_patient]] %>%
    arrange(desc(abs(stat))) %>%
    pull(gene) %>%
    str_subset(., ".+-AS[0-9]", negate = T) %>% # clear novel transcripts
    str_subset(., "^LINC[0-9]+", negate = T) %>%
    str_subset(., "^A(C|L|P)[0-9]+", negate = T)
  
  HT.list[[current_patient]] <- sce.MCL[, sce.MCL$patient == current_patient] %>%
    plotGroupedHeatmap(
      ., main = current_patient,
      features = genes.list[[current_patient]],
      group = "CNV.clone",
      center = TRUE,
      zlim = c(-0.1, 0.1),
      cluster_rows = T,
      angle_col = c("0"),
      silent = T, 
      fontsize_row = 6
    ) %>%
    ggplotify::as.ggplot()
}
```

```{r, fig.width=5, fig.height=10}
HT.list
```

```{r}
DECENT.GSE.list %>% 
  lapply(., as.data.frame)
```

```{r}
DECENT.GSE.list[["P009"]] %>% 
  as.data.frame() %>% 
  arrange(desc(NES)) %>% 
  slice_head(n = 20)
  
  # выбрать топ N путей (отсортировать по abs(NES))
  # и таблички с них сразу сгенерир
```


```{r, fig.width=8, fig.height=6}
NESdotplot <- function(DECENT.GSE.list, name, n_pathways) {
  pathways <- DECENT.GSE.list[[name]] %>%
    as.data.frame() %>%
    arrange(desc(NES)) %>%
    slice_head(n = n_pathways) %>%
    pull(Description)
  
  suppressMessages(
    DECENT.GSE.list[[name]] %>%
      ReactomePA::dotplot(., showCategory = pathways) +
      scale_x_continuous(labels = scales::percent) +
      geom_point(aes(fill = NES)) +
      scale_fill_gradient2(
        name = "NES",
        low = "blue",
        high = "red"
      )
  )
}

dotplots.list <- lapply(patients[1:4], function(x) NESdotplot(DECENT.GSE.list, x, 8))
dotplots.list
```

```{r, fig.width=9, fig.height=10}
HT.dotplot.list <- list()
for(i in seq_along(dotplots.list)){
  HT <- HT.list[[i]]
  dotplot <- dotplots.list[[i]]
  HT.dotplot.list[[patients[i]]] <- ggpubr::ggarrange(plotlist = list(HT, dotplot), nrow = 1)
}
HT.dotplot.list
```

```{r}
for(i in seq_along(HT.dotplot.list)){
  tiff(
    filename = file.path(figures.out.dir, "subclones_DE", paste0(patients[i], "_subcloneDE.tiff")),
    units = "in",
    width = 10,
    height = 12,
    res = 300,
    bg = "white"
  )
  print(HT.dotplot.list[[i]])
  dev.off()
}
```

### Save the tables

```{r}
clone.de.list %>% 
  lapply(., mutate,
         sig = case_when(
           pvalue < sig.pvalue & logfc > sig.logfc ~ "up",
           pvalue < sig.pvalue & logfc < -sig.logfc ~ "down",
           TRUE ~ "ns")
    ) %>% 
  lapply(., mutate, sig = factor(sig, levels = c("up", "down", "ns"))) %>% 
  lapply(arrange, sig) %>% 
  write.xlsx(file.path(tables.out.dir, "DECENT.DEG.by_clone.xlsx"))
```


```{r}
DECENT.GSE.list %>% 
  lapply(as.data.frame) %>% 
  lapply(remove_rownames) %>% 
  bind_rows(.id = "Patient") %>% 
  write.xlsx(file.path(tables.out.dir, "DECENT.GSE.RPA.by_clone.xlsx"))
```



