---
title: "1. Loading and filtering empty drops"
output: 
  html_notebook:
    toc: true
    toc_float: true
    code_folding: show
    theme: united
---

```{r setup}
library(tidyverse)
library(openxlsx)

# SC methods
library(DropletUtils)
library(Seurat)
library(scran)
library(scater)
library(bluster)
# Doublet scoring
library(scDblFinder)
# Annotations
library(SingleR)

# Variable with random seed value used throughout the analysis for reproducibility
random_seed_value = 17

# Parallelization
library(BiocParallel)
bpp <- MulticoreParam(workers = 60)

# Path to project folder
wd <- here::here()
```

# Load datasets

```{r}
patients <- c("P009", "P022", "P027", "P087", "P069")
timepoints <- c("DG", "REL")
compartments <- c("PBMC", "BM", "LN", "TON", "GUT", "ASC")

# Add the normal sample? We don't have it for this pipeline so don't.
add_normal = FALSE

if (add_normal){
  patients = c("N", patients)
  timepoints = c("N", timepoints)
  compartments = c("N", compartments)
}

input.path <- file.path(wd, "scRNAseq/alignment/output/")

# Construct metadata labels
samples.df <-
  list.dirs(input.path,
            full.names = F,
            recursive = F) %>%
  data.frame("sample" = .) %>%
  mutate(
    patient = word(sample, 1, sep = "_"),
    timepoint = word(sample, 2, sep = "_"),
    compartment = word(sample, 3, sep = "_"),
    repl = word(sample, 4, sep = "_") %>% str_remove(., "rep")
  ) %>%
  mutate(
    patient = factor(patient, levels = patients),
    timepoint = factor(timepoint, levels = timepoints),
    compartment = factor(compartment, levels = compartments)
  ) %>%
  arrange(patient, timepoint, compartment) %>%
  mutate(sample = factor(sample, levels = unique(sample)))

# Expected cells for the samples in the project:

samples.df <- data.frame(
  patient = patients,
  n.expected.cells = c(3500, 5000, 5000, 3500, 10000)
) %>% 
  left_join(
    samples.df, ., by = "patient"
  )

samples.df$patient <- factor(samples.df$patient, levels = patients)
samples.df
```

```{r}
sce_list <- list()
for(sample_name in samples.df$sample){
  sample_dir <- paste0(sample_name, "_Solo.out")
  
  sample.path <- file.path(input.path, sample_name, sample_dir, "Gene/raw/")
  
  sce_list[[sample_name]] <- DropletUtils::read10xCounts(
    sample.path, sample.names = sample_name, col.names = TRUE,
    BPPARAM = bpp)
  
  # Assign metadata labels
  sce_list[[sample_name]]$patient <- samples.df[which(samples.df$sample == sample_name), "patient"]
  sce_list[[sample_name]]$timepoint <- samples.df[which(samples.df$sample == sample_name), "timepoint"]
  sce_list[[sample_name]]$compartment <- samples.df[which(samples.df$sample == sample_name), "compartment"]
  
  # make colnames unique
  colnames(sce_list[[sample_name]]) <- paste(colData(sce_list[[sample_name]])[["Barcode"]],
                                             colData(sce_list[[sample_name]])[["Sample"]],
                                             sep = "-")
}
sce_list
```

# Droplets to cells

```{r}
for(sce_name in names(sce_list)) {
  sce <- sce_list[[sce_name]]
  
  libSizes <- colSums(assays(sce)$counts)
  
  sce_list[[sce_name]] <- sce[, libSizes > 0]
}

sce_list
```

```{r}
for(sce_name in names(sce_list)) {
  sce <- sce_list[[sce_name]]
  
  expected_cells <- samples.df[["n.expected.cells"]][which(samples.df$sample == sce_name)]
  
  set.seed(random_seed_value)
  res <- DropletUtils::emptyDropsCellRanger(counts(sce),
                                            n.expected.cells = expected_cells,
                                            niters = 50000,
                                            BPPARAM = bpp)
  
  is.cell <- which(res$Total > 100 & res$FDR <= 0.001)
  sce <- sce[, is.cell]
  
  sce_list[[sce_name]] <- sce
}

sce_list
```

# QC plots

```{r}
samples.df$total.cells <- NA
samples.df$discarded.lowQC.cells <- NA
for(sample.name in names(sce_list)) {
  sce <- sce_list[[sample.name]]
  
  QC.genes <- list(
    Mito = stringr::str_which(rowData(sce)[["Symbol"]], "^MT-"),
    Ribo = stringr::str_which(rowData(sce)[["Symbol"]], "^RPL")
    )
  
  sce <- scuttle::addPerCellQCMetrics(sce, subsets = QC.genes)
  qc.filter <- scuttle::quickPerCellQC(colData(sce),
                              sub.fields = c("subsets_Mito_percent"))
  colData(sce) <- cbind(colData(sce), qc.filter)
  
  samples.df[which(samples.df == sample.name), "total.cells"] <- dim(sce)[[2]]
  samples.df[which(samples.df == sample.name), "discarded.lowQC.cells"] <- table(sce$discard)[["TRUE"]]
  
  sce_list[[sample.name]] <- sce
}
```

```{r fig.height=10, fig.width=10, message=FALSE, warning=FALSE}
plots_genes_vs_sum <- list()
for (sample.name in names(sce_list)) {
  plots_genes_vs_sum[[sample.name]] <- plotColData(
    sce_list[[sample.name]],
    y = "sum",
    x = "detected",
    color_by = "discard",
    point_size = 1
  ) +
    scale_x_log10() +
    scale_y_log10() +
    scale_color_manual(values = c("gray30", "red"),
                       labels = c("Pass", "Low quality")) +
    labs(
      title = sample.name,
      color = "QC",
      x = "Detected unique genes",
      y = "Sum of transcript counts"
    )
}

ggpubr::ggarrange(
  plotlist = plots_genes_vs_sum,
  common.legend = T,
  ncol = 4,
  nrow = 4
)
```

```{r fig.height=10, fig.width=10, message=FALSE, warning=FALSE}
plots_genes_hist <- list()
for (sample.name in names(sce_list)) {
  plots_genes_hist[[sample.name]] <- ggcells(sce_list[[sample.name]],
                                             aes(x = detected)) +
    geom_histogram(aes(fill = discard),
                   binwidth = 100) +
    theme_bw() +
    scale_fill_manual(values = c("gray30", "red"),
                      labels = c("Pass", "Low quality")) +
    labs(
      title = sample.name,
      fill = "QC",
      x = "Detected unique genes",
      y = "Cells"
    )
}

ggpubr::ggarrange(
  plotlist = plots_genes_hist,
  common.legend = T,
  ncol = 4,
  nrow = 4
)
```

```{r fig.height=10, fig.width=10, message=FALSE, warning=FALSE}
plots_sum_vs_mito <- list()
for (sample.name in names(sce_list)) {
  plots_sum_vs_mito[[sample.name]] <- ggcells(sce_list[[sample.name]],
                                              aes(x = sum,
                                                  y = subsets_Mito_percent)) +
    geom_point(aes(color = discard),
               size = 1, alpha = 0.75) +
    geom_density_2d(color = "blue") +
    cowplot::theme_cowplot() +
    scale_color_manual(values = c("gray30", "red"),
                       labels = c("Pass", "Low quality")) +
    scale_x_log10() +
    labs(
      title = sample.name,
      fill = "QC",
      x = "Sum of transcript counts",
      y = "% of mitochondrial genes"
    )
}

ggpubr::ggarrange(
  plotlist = plots_sum_vs_mito,
  common.legend = T,
  ncol = 4,
  nrow = 4
)
```

# Doublet detection

```{r}
samples.df$discarded.doublet.cells <- NA
doublet.plots <- list()
for (sample.name in names(sce_list)) {
  sce <- sce_list[[sample.name]]
  
  set.seed(random_seed_value)
  clusters_igraph <- quickCluster(sce,
                                  method = "igraph",
                                  BPPARAM = bpp)
  sce$clusters_igraph <- clusters_igraph
  
  set.seed(random_seed_value)
  doublet_scores <- scDblFinder(sce,
                                clusters = factor(sce$clusters_igraph),
                                returnType = "table")
  set.seed(random_seed_value)
  doublet_calls <-
    doubletThresholding(doublet_scores, method = "griffiths",
                        returnType = "call")
  
  # Assign doublet counts
  sce$scDblFinder.calls <- doublet_calls
  
  samples.df[which(samples.df == sample.name), "discarded.doublet.cells"] <- table(sce$scDblFinder.calls)[["doublet"]]
  
  sce_list[[sample.name]] <- sce
}
samples.df
```

```{r fig.height=10, fig.width=10, message=FALSE, warning=FALSE}
plots_doublets <- list()
for (sample.name in names(sce_list)) {
  plots_doublets[[sample.name]] <- plotColData(
    sce_list[[sample.name]],
    y = "sum",
    x = "detected",
    color_by = "scDblFinder.calls",
    point_size = 1
  ) +
    scale_x_log10() +
    scale_y_log10() +
    scale_color_manual(values = c("gray30", "red"),
                       labels = c("Singlet", "Doublet")) +
    labs(
      title = sample.name,
      color = "QC",
      x = "# of detected genes",
      y = "Sum of transcript counts"
    )
}
ggpubr::ggarrange(
  plotlist = plots_doublets,
  common.legend = T,
  ncol = 4,
  nrow = 4
)
```

# Join SCE list into one annotated SCE object

```{r}
joined_sce <- do.call(cbind, sce_list)
names(metadata(joined_sce)) <- unique(joined_sce$Sample)
joined_sce[["Sample"]] <- factor(joined_sce[["Sample"]], levels = unique(joined_sce$Sample))
joined_sce[["patient"]] <- factor(joined_sce[["patient"]], levels = patients)
colData(joined_sce) %>% as.data.frame() %>% head()
```

# Final cell count

```{r}
samples.df <- samples.df %>% 
  mutate(
    QC.cells = total.cells - (discarded.lowQC.cells + discarded.doublet.cells)
  )
samples.df
```

# Store the SCE

```{r}
iteration_output = "0_cells"
storage.dir <- file.path(wd, "scRNAseq/analysis/outputs/", iteration_output)
dir.create(storage.dir, showWarnings = F, recursive = T)
saveRDS(joined_sce,
        file = paste(storage.dir, "sce.rds", sep = "/")
        )
write.xlsx(samples.df,
           paste(storage.dir, "cell_metadata.xlsx", sep = "/")
           )
```

```{r}
sessionInfo()
```

